design/jira-traceability.md

```diff
@@ -0,0 +1,150 @@
+4. Filters (E.g. branch filters) to restrict the information transfer to JIRA.
+
+
+## Interaction with Jira Cloud 
+
+### Pull:
+ For pulling information from JIRA we would just need the service connection of the JIRA server. Using this we can pull details of consumed work items and display the list in  azure pipelines. 
+
+### Push:
+Jira currently supports REST APIs that let external systems push build and deployment details in JIRA. There are 2 separate APIs for [build](https://developer.atlassian.com/cloud/jira/software/rest/#api-rest-builds-0-1-bulk-post) and [deployments](https://developer.atlassian.com/cloud/jira/software/rest/#api-rest-deployments-0-1-bulk-post). While JIRA depends on external services to fill these data, the work item UI is entirely painted by JIRA only. External services do not have any control over the UI.
+
+#### Jira Build UI and APIs:
+The  build results are populated in the development section of the work item UI. On further clicking on it, we see the pipeline runs that were associated with the build. 
+
+![Build1](./images/jira-traceability-build-1.png)
+
+
+Drill down view:
+![Build2](./images/jira-traceability-build-2.png)
+
+For the API, the combination of `pipelineId` (pipeline definition identifier) and `buildNumber` (pipeline run identifier) is considered as the unique identifier for a build. 
+
+#### Jira Deployment UI and APIs:
+
+Jira has a separate section for Deployments in its work item UI. In this, only the highest deployed environment category is displayed. Eg. If the work item was deployed to a `test` type environment and  then to `production` type, Jira shows only `production` in deployment section. On clicking on this entry the user can see details of previous deployments.
+
+![Release1](./images/jira-traceability-deploy-1.png)
+
+Drill down view:
+![Release2](./images/jira-traceability-deploy-2.png)
+
+For the API, the combination of `pipelineId` (pipeline definition identifier), `environmentID` (environment identifier), `deploymentSequenceNumber` (deployment identifier) constitute a unique identifier for deployment. 
+
+
+#### When should the data be pushed from Azure Pipelines:?
+ Based on customer interactions, customers prefer seeing the build status as soon as a build completes and deployment status when deployment to an environment is complete. Since build generation and deployments are done at a job level in azure pipelines, we'll push the status information to Jira at job boundaries. For deployments, azure pipelines already has a separate category of jobs, viz, deployment jobs. We'll invoke deployment APIs at the beginning and end of deployment jobs if all the filter conditions are met. For builds, however, we currently do not have a separate category of jobs. For this, we will assume that any job that publishes an artifact is a build job. We'll invoke build API at the end of these jobs if filter conditions are met.  
```

Comment:
`we will assume that any job that publishes an artifact is a build job` is this means, if a job is not publishing anything then we will not push the linking?

Reply:
You're right. We'll assume that any non-deployment job that publishes is a build job. We will push information for such build jobs only.